From c1b21052dc8607c71b75c680a680be86cc5d99a1 Mon Sep 17 00:00:00 2001
From: wfeng6 <wei.feng@intel.com>
Date: Thu, 18 Oct 2012 21:15:32 +0800
Subject: Accelerate the thumbnail extraction when the sync-frame duration is long

BZ: 58036

When the sync frame duration is very long in the test clips, the
thumbnail extration will be very slow. Because the program will
have to decode from the previous sync-frame to the desired frame.
If end user presses the exit button before the thumbnail
extraction finishes, sometimes ANR will happen.

In this patch, if the desized frame's timestamp is 3secs larger
than the previous sync-frame, we will choose to use the sync-frame
as the thumbnail.

Category:aosp improvement
Domain: Video.Media-video editor
Origin: internal
Upstream-Candidate: No

Change-Id: I95cbcb80005d5482ef70c3d4de48fd3c96002036
Orig-Change-Id: I0add774a1d56837ad14d1cc089965d64e757dc00
Signed-off-by: wfeng6 <wei.feng@intel.com>
Signed-off-by: Gu, Wangyi <wangyi.gu@intel.com>
Signed-off-by: chuzhoux <chux.zhou@intel.com>
---
 libvideoeditor/vss/common/inc/M4DECODER_Common.h   |    5 +++++
 .../inc/VideoEditorVideoDecoder_internal.h         |    2 ++
 2 files changed, 7 insertions(+), 0 deletions(-)

diff --git a/libvideoeditor/vss/common/inc/M4DECODER_Common.h b/libvideoeditor/vss/common/inc/M4DECODER_Common.h
index 93e3062..e4ac77a 100755
--- a/libvideoeditor/vss/common/inc/M4DECODER_Common.h
+++ b/libvideoeditor/vss/common/inc/M4DECODER_Common.h
@@ -135,6 +135,11 @@ typedef enum
     M4DECODER_kOptionID_VideoDecodersAndCapabilities =
         M4OSA_OPTION_ID_CREATE(M4_READ, M4DECODER_COMMON, 0x10),
 
+    /**
+     * Set the mode of the video decoder */
+    M4DECODER_kOptionID_VideoDecodersThumbnailMode =
+        M4OSA_OPTION_ID_CREATE(M4_READ, M4DECODER_COMMON, 0x20),
+
     /* common to MPEG4 decoders */
     /**
      * Get the DecoderConfigInfo */
diff --git a/libvideoeditor/vss/stagefrightshells/inc/VideoEditorVideoDecoder_internal.h b/libvideoeditor/vss/stagefrightshells/inc/VideoEditorVideoDecoder_internal.h
index 6762643..ae3a894 100755
--- a/libvideoeditor/vss/stagefrightshells/inc/VideoEditorVideoDecoder_internal.h
+++ b/libvideoeditor/vss/stagefrightshells/inc/VideoEditorVideoDecoder_internal.h
@@ -118,6 +118,8 @@ typedef struct {
     // Time interval between two consequtive/neighboring video frames.
     M4_MediaTime            mFrameIntervalMs;
 
+    // Indicate whether the decoder is used to generate the thumbnail.
+    M4OSA_Int32             mThumbnail;
 } VideoEditorVideoDecoder_Context;
 
 } //namespace android
-- 
1.7.4.1

